### ISSUE TITLE
How the padding is done for depthwise convolution?

### ISSUE THREAD

**Issue by benjamim93** [oneapi-src/oneDNN#787]
Could you please give an explanation on how padding is done for depthwise convolution (maybe show me where the related part is in the code)?  I see there is some calculation of the actual right and bottom padding sizes, then what is the meaning of passing these values through the memory descriptor?

**Comment by vpirogov** [comment]
Hi @benjamim93,

Padding for convolution spatial dimensions is specified in convolution descriptor. You can find reference implementation of convolution [here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/ref_convolution.cpp#L66) and formulas [here](https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html). It maps directly to how padding is specified in Tensorflow or Pytorch.

Padding in memory descriptor specifies physical layout of the data in the buffer to handle situations when oneDNN needs to pad channel dimension with zeroes for performance reasons. It does not affect logical operation. In case you are interested you can find more details on oneDNN memory formats [here](https://oneapi-src.github.io/oneDNN/dev_guide_understanding_memory_formats.html).

**Comment by emfomenk** [comment]
> I see there is some calculation of the actual right and bottom padding sizes, then what is the meaning of passing these values through the memory descriptor?

@benjamim93, could you please point where you see these "_calculation of the actual right and bottom padding sizes_"?

**Comment by benjamim93** [comment]
Hi @emfomenk, I see in /src/cpu/x64/jit_uni_dw_conv_kernel_utils.hpp ln 208-211:

` jcp.r_pad = calculate_end_padding(jcp.l_pad, jcp.ow, jcp.iw, jcp.stride_w, ext_kw);`
` jcp.b_pad = calculate_end_padding(jcp.t_pad, jcp.oh, jcp.ih, jcp.stride_h, ext_kh);`

**Comment by emfomenk** [comment]
Hi @benjamim93,

The code here is used inside the library for the actual implementation. It computes how much padding on the right (or bottom) the compute-kernel should be aware of. This is the lower bound, which doesn't necessarily equal to the right (or bottom) padding you provided in the API. Example:

- Case 1:
    ```
    inputs width:  IW = 1
    kernel width:  KW = 2
    stride:        SW = 2
    left padding:  PW_L = 0
    right padding: PW_R = 1

    --> output width = 1
    ```
- Case 2:
    ```
    inputs width:  IW = 1
    kernel width:  KW = 2
    stride:        SW = 2
    left padding:  PW_L = 0
    right padding: PW_R = 2 // before it was 1

    --> output width = 1 // and the whole output is the same as in case 1
    ```

Here, you can see that output-wise the results are absolutely the same for both cases. The oneDNN API gives you some flexibility on how you choose the right (and bottom) paddings, and indeed that could be a little bit redundant.
So, the code you reference above solves this _redundancy_ for the implementation: for both cases the `jcp.r_pad` will be equal to 1.

But this is **an implementation detail** (that we did to simplify the code inside the library). From a user perspective the only important thing is what is passed through the memory descriptors and convolution op descriptor parameters. The result should comply to the definition given in [here](https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html).

**Comment by benjamim93** [comment]
Thanks @emfomenk @vpirogov
